rfcs/0000-structured-logging.md

```diff
@@ -0,0 +1,2189 @@
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Design considerations
+
+### Don't break anything
+
+Allow structured logging to be added in the current `0.4.x` series of `log`. This gives us the option of including structured logging in an `0.4.x` release, or bumping the minor crate version without introducing any actual breaking changes.
+
+### Don't create a public dependency
+
+Don't create a new serialization framework that causes `log` to become a public dependency of any library that wants their data types to be loggable. Logging is a truly cross-cutting concern, so if `log` was a public dependency it would probably be at least as pervasive as `serde` is now.
+
+### Object safety
+
+`log` is already designed to be object-safe so this new structured logging API needs to be object-safe too.
+
+### Borrowed vs owned
+
+`Record` borrows all data from the call-site so records need to be handled directly on-thread as they're produced. On the one hand that means that log records need to be serialized before they can be sent across threads. On the other hand it means callers don't need to make assumptions about whether records need to be owned or borrowed.
+
+A scheme for more efficient conversion from borrowed to owned sources of key-value pairs can be added later.
+
+## Cargo features
+
+Structured logging will be supported in either `std` or `no_std` contexts using Cargo features:
+
+```toml
+[features]
+# optional serde-support requires `std`
+kv_serde = ["std", "serde", "erased-serde"]
+```
+
+Using default features, structured logging will be supported by `log` in `no_std` environments for a fixed set of types from the standard library. Using the `kv_serde` feature, any type that implements `Debug + Serialize` can be logged, and its potentially complex structure will be retained.
+
+## Implications for dependents
+
+Dependents of `log` will notice the following:
+
+### Default crate features
+
+The API that's available with default features doesn't add any extra dependencies to the `log` crate, and shouldn't impact compile times or artifact size much.
+
+### After opting in to `kv_serde`
+
+In `no_std` environments (which is the default for `log`):
+
+- `serde` will enter the `Cargo.lock` if it wasn't there already. This will impact compile-times.
+- Artifact size of `log` will increase.
+
+In `std` environments (which is common when using `env_logger` and other crates that implement `Log`):
+
+- `serde` and `erased-serde` will enter the `Cargo.lock` if it wasn't there already. This will impact compile-times.
+- Artifact size of `log` will increase.
+
+In either case, `serde` will become a public dependency of the `log` crate, so any breaking changes to `serde` will result in breaking changes to `log`.
+
+## Public API
+
+For context, ignoring the `log!` macros, this is roughly the additional public API this RFC proposes to support structured logging:
+
+```rust
```

Comment:
This entire code section is repeated (from what I can tell) below in different sections effectively making this bit obsolete and only making the RFC longer.